observation/interferometric_pointing.py

```diff
@@ -57,48 +57,49 @@ class NoTargetsUpError(Exception):
 
         start_time = time.time()
         targets_observed = []
-        # Keep going until the time is up
-        keep_going = True
-        while keep_going:
-            keep_going = (opts.max_duration is not None)
-            targets_before_loop = len(targets_observed)
-            # Iterate through source list, picking the next one that is up
-            for target in observation_sources.iterfilter(el_limit_deg=opts.horizon):
-                # Check if all offset pointings in compound scan will be up
-                compound_steps = 1 + 2 * (opts.number_of_steps // 2)
-                # Add some extra time for slews between pointings
-                step_duration = opts.track_duration + 4.
-                compound_duration = compound_steps * step_duration
-                if not session.target_visible(target, duration=compound_duration):
-                    continue
-                session.label('interferometric_pointing')
-                session.track(target, duration=opts.track_duration, announce=False)
-                for direction in {'x', 'y'}:
-                    for offset in np.linspace(-opts.max_extent, opts.max_extent,
-                                              opts.number_of_steps // 2):
-                        if direction == 'x':
-                            offset_target = [offset, 0.0]
-                        else:
-                            offset_target = [0.0, offset]
-                        user_logger.info("Initiating %g-second track on target '%s'",
-                                         opts.track_duration, target.name)
-                        user_logger.info("Offset of (%f, %f) degrees", *offset_target)
-                        if not kat.dry_run:
-                            session.ants.req.offset_fixed(offset_target[0], offset_target[1], opts.projection)
-                        nd_params = session.nd_params
-                        #session.fire_noise_diode(announce=True, **nd_params)
-                        target.tags = target.tags[:1]  # this is to avoid overloading the cal pipeline
-                        session.track(target, duration=opts.track_duration, announce=False)
-                targets_observed.append(target.name)
-                if opts.max_duration is not None and (time.time() - start_time >= opts.max_duration):
-                    user_logger.warning("Maximum duration of %g seconds has elapsed - stopping script",
-                                        opts.max_duration)
-                    keep_going = False
-                    break
+        try: # Keep going until the time is up or interrupted
```

Comment:
I'm not a big fan of massive `try...finally` blocks (in Python, indentation is considered harmful in my books). One option would be to register something with `session.end()` (more work). A better solution would be to add your own context manager to the `with start_session(...) as session:` block, doing something like:

```Python
from contextlib import contextmanager

targets_observed = []

@contextmanager
def report_targets():
    try:
        yield
    finally:
        user_logger.info("Targets observed : %d (%d unique)",
                                    len(targets_observed), len(set(targets_observed)))

with start_session() as session, report_targets():
   # ...
```

Reply:
it seems like i am making spaghetti of the review comment threads...

my latest commit implements the suggestion by @spassmoor and @ludwigschwardt to scan along a circle around bore sight. it also happens to remove the deepest nesting from the code and also shortens the nested code by 5 lines.